langchain4j/src/main/java/dev/langchain4j/data/document/source/ClassPathSource.java

```diff
@@ -0,0 +1,97 @@
+package dev.langchain4j.data.document.source;
+
+import static dev.langchain4j.data.document.Document.FILE_NAME;
+import static dev.langchain4j.data.document.Document.URL;
+import static dev.langchain4j.internal.ValidationUtils.ensureNotBlank;
+import static dev.langchain4j.internal.ValidationUtils.ensureNotNull;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+
+import dev.langchain4j.data.document.DocumentSource;
+import dev.langchain4j.data.document.Metadata;
+
+/**
+ * Specialization of a {@link DocumentSource} that knows how to read from the classpath
+ */
+public class ClassPathSource implements DocumentSource {
+    public static final String IS_INSIDE_ARCHIVE = "is_inside_archive";
+    private final URL url;
+    private final ClassLoader classLoader;
+    private final Metadata metadata = new Metadata();
+
+    public ClassPathSource(URL url, ClassLoader classLoader) {
+        this.url = ensureNotNull(url, "url");
+        this.classLoader = classLoader;
+
+        var file = this.url.getFile();
+        this.metadata.put(URL, file);
+        this.metadata.put(FILE_NAME, file.substring(file.lastIndexOf(File.separatorChar) + 1));
+        this.metadata.put(IS_INSIDE_ARCHIVE, "jar".equalsIgnoreCase(this.url.getProtocol()));
```

Comment:
I am not sure we should keep this in the metadata. Why is this metadata entry required? AFAIS, it is used only in the tests?

Reply:
I don't have a strong opinion. There wasn't a specific "need" for it, but I just thought it would be nice to know whether or not it was in an archive. If I get rid of this I can also get rid of the changes I made to `Metadata`.

Reply:
If there is no specific need, let's better remove it please?